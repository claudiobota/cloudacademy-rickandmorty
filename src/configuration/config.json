{
	"api": {
		"url": "https://rickandmortyapi.com/api",
		"paths": {
			"characters": "/character",
			"locations": "/location",
			"episodes": "/episode"
		},
		"page_size": 20
	}
}
