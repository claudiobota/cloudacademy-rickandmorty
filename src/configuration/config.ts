import configData from './config.json';
import configDevelopmentData from './config.development.json';
import configTestData from './config.test.json';
import configStagingData from './config.staging.json';
import configProductionData from './config.production.json';
import {IConfig} from '../types/interfaces';
import {Environment} from '../types/enums';

let config: IConfig = {
  ...configData
};

const nodeEnv: string = process.env.NODE_ENV || 'development';

switch (nodeEnv) {
  case Environment.DEVELOPMENT:
    config = {
      ...config,
      ...configDevelopmentData
    };
    break;

  case Environment.TEST:
    config = {
      ...config,
      ...configTestData
    };
    break;

  case Environment.STAGING:
    config = {
      ...config,
      ...configStagingData
    };
    break;

  case Environment.PRODUCTION:
    config = {
      ...config,
      ...configProductionData
    };
    break;

  default:
}

export default config;
