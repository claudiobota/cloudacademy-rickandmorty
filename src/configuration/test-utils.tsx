import React from 'react';
import {render as rtlRender} from '@testing-library/react';
import {Provider} from 'react-redux';
import {configureAppStore} from '../redux/store';

function render(
  ui: React.ReactElement,
  {
    initialState,
    store = configureAppStore(initialState),
    ...renderOptions
  }: any = {}
) {
  function Wrapper({ children }: { children: React.ReactElement }) {
    return <Provider store={store}>{children}</Provider>;
  }

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { render }
