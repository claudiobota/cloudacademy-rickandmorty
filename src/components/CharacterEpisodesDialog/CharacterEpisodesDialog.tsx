import React, {forwardRef, useEffect, useImperativeHandle, useState} from 'react';
import {Dialog} from 'primereact/dialog';
import {ICharacter, IEpisode} from '../../types/interfaces';
import {DataTable} from 'primereact/datatable';
import {Column} from 'primereact/column';
import {useSelector} from 'react-redux';
import {selectAllEpisodes} from '../../redux/slices/episodeSlice';
import config from '../../configuration/config';

interface IProps {
  character: ICharacter;
}

function CharacterEpisodesDialog(props: IProps, ref: any) {
  const [isVisible, setVisible] = useState(false);
  const [characterEpisodes, setCharacterEpisodes] = useState<IEpisode[]>([]);
  const episodes = useSelector(selectAllEpisodes);

  useEffect(() => {
    const episodeIds = props.character.episode.map(episodeUrl => Number(episodeUrl.replace(`${config.api.url}${config.api.paths.episodes}/`, '')));
    setCharacterEpisodes(episodes.filter(episode => episodeIds.indexOf(episode.id) > -1));
  }, [episodes]);

  useImperativeHandle(
    ref,
    () => ({
      show() {
        setVisible(true);
      }
    })
  );

  const hideDialog = () => {
    setVisible(false);
  };

  return (
    <Dialog
      header={`View all episodes of ${props.character.name}`}
      visible={isVisible}
      style={{ width: '95vw', maxWidth: '800px' }}
      className=""
      onHide={() => hideDialog()}
    >
      <DataTable value={characterEpisodes} className="p-datatable-gridlines">
        <Column field="name" header="Name" />
        <Column field="air_date" header="Air Date" />
        <Column field="episode" header="Episode" />
      </DataTable>
    </Dialog>
  );
}

export default forwardRef(CharacterEpisodesDialog);
