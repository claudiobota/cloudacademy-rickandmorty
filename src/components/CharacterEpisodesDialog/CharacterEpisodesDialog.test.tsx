import React from 'react';
import {render} from '../../configuration/test-utils';
import CharacterEpisodesDialog from './CharacterEpisodesDialog';
import mockData from '../../configuration/mock-data.json';
import {act} from '@testing-library/react';

describe('CharacterEpisodesDialog', () => {
  test('renders and matches snapshot', () => {
    let {asFragment} = render(<CharacterEpisodesDialog character={mockData.api.characters.results[0]} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders visible and matches snapshot',() => {
    const ref: any = React.createRef();
    const component = render(<CharacterEpisodesDialog ref={ref} character={mockData.api.characters.results[0]} />);

    act(() => {
      ref.current.show();
    });

    expect(component.asFragment()).toMatchSnapshot();
  });
});
