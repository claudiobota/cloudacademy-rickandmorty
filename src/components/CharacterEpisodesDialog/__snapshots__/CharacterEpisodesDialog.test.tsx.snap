// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterEpisodesDialog renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-dialog-mask p-component-overlay p-dialog-center"
  />
</DocumentFragment>
`;

exports[`CharacterEpisodesDialog renders visible and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-dialog-mask p-component-overlay p-dialog-visible p-dialog-center"
    style="z-index: 1000;"
  >
    <div
      aria-labelledby="pr_id_2_header"
      class="p-dialog p-component p-dialog-enter p-dialog-enter-active"
      id="pr_id_2"
      role="dialog"
      style="width: 95vw; max-width: 800px;"
    >
      <div
        class="p-dialog-header"
      >
        <span
          class="p-dialog-title"
          id="pr_id_2_header"
        >
          View all episodes of Rick Sanchez
        </span>
        <div
          class="p-dialog-header-icons"
        >
          <button
            aria-label="Close"
            class="p-dialog-header-icon p-dialog-header-close p-link"
            type="button"
          >
            <span
              class="p-dialog-header-close-icon pi pi-times"
            />
          </button>
        </div>
      </div>
      <div
        class="p-dialog-content"
      >
        <div
          class="p-datatable p-component p-datatable-gridlines"
          data-scrollselectors=".p-datatable-scrollable-body, .p-datatable-unfrozen-view .p-datatable-scrollable-body"
        >
          <div
            class="p-datatable-wrapper"
          >
            <table
              role="grid"
            >
              <thead
                class="p-datatable-thead"
              >
                <tr>
                  <th
                    class=""
                  >
                    <span
                      class="p-column-title"
                    >
                      Name
                    </span>
                  </th>
                  <th
                    class=""
                  >
                    <span
                      class="p-column-title"
                    >
                      Air Date
                    </span>
                  </th>
                  <th
                    class=""
                  >
                    <span
                      class="p-column-title"
                    >
                      Episode
                    </span>
                  </th>
                </tr>
              </thead>
              <tbody
                class="p-datatable-tbody"
              />
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
